### ISSUE TITLE
Overflow errors give wrong position information when called via traits

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust#152599]
I tried this code:

```rust
#[allow(arithmetic_overflow)]
fn main() {
    let _ = core::ops::Shr::shr(1u8, 25usize);
}
```

I expected this to report the `main` function as the location of a panic due to an overflow error.

Instead, the panic used the location of the definition of `impl Shr<usize> for u8` as the location instead.

Further background: when calling the arithmetic operations directly, the actual line of the operation is reported as the location for the error. However, when this is done via traits, the trait itself is listed as the location of the error. You can force it directly by calling the trait method, however, the main use case is where the trait is "indirectly" used, e.g. by a method which is generic over arithmetic traits.

Note that since trait implementers can annotate methods with `#[track_caller]` independent of the original definition, I would expect the behaviour of this to be identical in the case of arithmetic overflows on these methods, however, that is not the case here. Currently, the only attribute there is `#[rustc_inherit_overflow_checks]`, which does not add this additional behaviour.

### Meta

`rustc --version --verbose`:
```
rustc 1.95.0-nightly (474276961 2026-01-26)
binary: rustc
commit-hash: 474276961f48b0d05f4ea260ba400096b027584e
commit-date: 2026-01-26
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8
```
